perf(forms): make `FormBuilder` and `RadioControlRegistry` tree-shakable

AndrewKushnir opened the pull request:
This PR makes the `FormBuilder` and `RadioControlRegistry` classes tree-shakable by adding the `providedIn` property to corresponding `@Injectable` decorators and excluding these classes from NgModule definitions.

Note: this improvement is a part of #41011.

## PR Type
What kind of change does this PR introduce?

- [x] Other... Please describe: performance optimization.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

AndrewKushnir committed:
perf(forms): make `FormBuilder` class tree-shakable

This commit makes the `FormBuilder` class tree-shakable by adding the `providedIn` property to its `@Injectable`
decorator. Now if the `FormBuilder` class is not referenced in application's code, it should not be included into
its production bundle.

AndrewKushnir commented:
[Presubmit](http://test/OCL:361708725:BASE:361708740:1615258561436:6a9f3e38).

mary-poppins commented:
You can preview 5482ea3 at https://pr41126-5482ea3.ngbuilds.io/.
You can preview e19a9dd at https://pr41126-e19a9dd.ngbuilds.io/.

AndrewKushnir committed:
perf(forms): make `RadioControlRegistry` class tree-shakable

This commit makes the `RadioControlRegistry` class tree-shakable by adding the `providedIn` property to its
`@Injectable` decorator. Now if the radio buttons are not used in the app (thus no `RadioControlValueAccessor`
directive is initialized), the `RadioControlRegistry` should not be included into application's prod bundle.

mary-poppins commented:
You can preview ef68a26 at https://pr41126-ef68a26.ngbuilds.io/.

mary-poppins commented:
You can preview acb48ff at https://pr41126-acb48ff.ngbuilds.io/.

mary-poppins commented:
You can preview 419bbe6 at https://pr41126-419bbe6.ngbuilds.io/.

AndrewKushnir commented:
Note: this change should land after #41146.

mary-poppins commented:
You can preview d35d95a at https://pr41126-d35d95a.ngbuilds.io/.

mary-poppins commented:
You can preview d0d2d59 at https://pr41126-d0d2d59.ngbuilds.io/.

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:361708725:BASE:362739824:1615683284768:e9c4f8c8) + [Global Presubmit](https://test.corp.google.com/OCL:361708725:BASE:362739865:1615683346431:ade5dc47).

AndrewKushnir commented on packages/core/test/bundling/forms_reactive/bundle.golden_symbols.json:
```diff
@@ -450,7 +450,7 @@
   },
   {
-    "name": "RadioControlRegistry"
+    "name": "RadioControlRegistryModule"
```
This change indicates that the `RadioControlRegistry` was tree-shaken away, but the `RadioControlRegistryModule` shows up here as it's used as the `providedIn` value (see below). Note: the `FormBuilder` class is not showing up here since it's used in that demo app (`forms_reactive`) and the `forms_template_driven` golden file never contained it. I've verified that both symbols are not present in a simple app that don't use radio buttons and doesn't reference `FormsBuilder` directly.

petebacondarwin commented on packages/forms/src/directives/radio_control_value_accessor.ts:
```diff
@@ -24,11 +24,21 @@ function throwNameError() {
+ * Internal-only NgModule that works as a host for the `RadioControlRegistry` tree-shakable
+ * provider. Note: the `InternalFormsSharedModule` can not be used here directly, since it's
+ * declared *after* the `RadioControlRegistry` class and the `providedIn` doesn't support
+ * `forwardRef` logic.
```
Should we consider adding support for `ForwardRef` to `providedIn`?

AndrewKushnir replied:
Yeah, created a ticket #41205.

petebacondarwin commented on packages/forms/src/form_providers.ts:
```diff
@@ -9,21 +9,21 @@
  * Exports the required providers and directives for template-driven forms,
  * making them available for import by NgModules that import this module.
  *
+ * Providers associated with this module:
+ * * `RadioControlRegistry`
+ *
```
We need to update the doc-gen to be able to join up `providedIn` injectables with their NgModule. I'll create a separate issue to track that.

petebacondarwin replied:
https://github.com/angular/angular/issues/41203

AndrewKushnir replied:
Thanks Pete 👍

petebacondarwin reviewed (approved):
More simple but effective improvements @AndrewKushnir !

jessicajaniuk reviewed (approved):
LGTM 🍪 

reviewed-for: public-api

jelbourn reviewed (approved):
LGTM

Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api
